agent/engine/dockerclient/dockerclientfactory.go

```diff
@@ -0,0 +1,116 @@
+// Copyright 2014-2015 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//	http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package dockerclient
+
+import (
+	"sync"
+
+	"github.com/aws/amazon-ecs-agent/agent/engine/dockeriface"
+	log "github.com/cihub/seelog"
+	docker "github.com/fsouza/go-dockerclient"
+)
+
+type Factory interface {
+	// GetDefaultClient returns a versioned client for the default version
+	GetDefaultClient() (dockeriface.Client, error)
+
+	// GetClient returns a client with the specified version
+	GetClient(version DockerVersion) (dockeriface.Client, error)
+
+	// FindSupportedAPIVersions tests each agent supported API version
+	// against the docker daemon and returns a slice of supported
+	// versions. The slice will always be a subset (non-strict)
+	// of agent supported versions.
+	FindSupportedAPIVersions() []DockerVersion
+}
+
+type factory struct {
+	endpoint string
+	lock     sync.Mutex
+	clients  map[DockerVersion]dockeriface.Client
+}
+
+// newVersionedClient is a variable such that the implementation can be
+// swapped out for unit tests
+var newVersionedClient = func(endpoint, version string) (dockeriface.Client, error) {
+	return docker.NewVersionedClient(endpoint, version)
+}
+
+func NewFactory(endpoint string) Factory {
+	return &factory{
+		endpoint: endpoint,
+		clients:  make(map[DockerVersion]dockeriface.Client),
+	}
+}
+
+func (f *factory) GetDefaultClient() (dockeriface.Client, error) {
+	return f.GetClient(getDefaultVersion())
+}
+
+func (f *factory) GetClient(version DockerVersion) (dockeriface.Client, error) {
+	client, ok := f.clients[version]
+	if ok {
+		return client, nil
+	}
+
+	f.lock.Lock()
+	defer f.lock.Unlock()
+
+	// double-check now that we're in a lock
+	client, ok = f.clients[version]
+	if ok {
+		return client, nil
+	}
+
+	client, err := newVersionedClient(f.endpoint, string(version))
+	if err != nil {
+		return nil, err
+	}
+
+	err = client.Ping()
+	if err != nil {
+		return nil, err
+	}
+
+	f.clients[version] = client
+	return client, nil
+}
+
+func (f *factory) FindSupportedAPIVersions() []DockerVersion {
```

Comment:
Can you add a comment for this?

// FindSupportedAPIVersions returns the API versions which are both known to exist for the Docker daemon to which the agent is connected and are supported by the agent.